src/command_buffer.rs

```diff
@@ -0,0 +1,217 @@
+    /// let mut recorder = CommandBuffer::new();
+    /// recorder.spawn_at(a, (false,0.0));
+    /// recorder.run_on(&mut world);
+    /// assert!(world.contains(a));
+    /// ```
+    pub fn run_on(&mut self, world: &mut World) {
+        for index in self.entity.iter() {
+            self.info[index.begin..index.end].sort_unstable_by_key(|z| z.ty_info);
+        }
+
+        for _ in 0..self.entity.len() {
+            let (ent, comps) = self.build();
+            world.spawn_at(ent, comps);
+            self.mark += 1;
+        }
+        self.clear();
+    }
+
+    fn build(&mut self) -> (Entity, ReadyBuffer<'_>) {
+        // Amount of components entity has
+        let len = self.entity[self.mark].end - self.entity[self.mark].begin;
+        self.ids
+            .extend(self.info[0..len].iter().map(|x| x.ty_info.id()));
+        let entity = self.entity[self.mark].entity;
+        (entity, ReadyBuffer { buffer: self })
+    }
+
+    /// Drop previously `recorded` entities and their components
+    ///
+    /// Recorder is cleared implicitly when entities are spawned, so usually this doesn't need to
+    /// be called
+    pub fn clear(&mut self) {
+        self.ids.clear();
+        self.entity.clear();
+        self.cursor = 0;
+        self.mark = 0;
+        unsafe {
+            for info in self.info.drain(..) {
+                info.ty_info.drop(self.storage.as_ptr().add(info.offset));
+            }
+        }
+    }
+}
+
+unsafe impl Send for CommandBuffer {}
+unsafe impl Sync for CommandBuffer {}
+
+impl Drop for CommandBuffer {
+    fn drop(&mut self) {
+        self.clear();
+        if self.layout.size() != 0 {
+            unsafe {
+                dealloc(self.storage.as_ptr(), self.layout);
+            }
+        }
+    }
+}
+
+impl Default for CommandBuffer {
+    /// Create an empty buffer
+    fn default() -> Self {
+        Self {
+            entity: Vec::new(),
+            storage: NonNull::dangling(),
+            layout: Layout::from_size_align(0, 8).unwrap(),
+            cursor: 0,
+            info: Vec::new(),
+            ids: Vec::new(),
+            mark: 0,
+        }
+    }
+}
+
+/// The output of an '[CommandBuffer]` suitable for passing to
+/// [`World::spawn_into`](crate::World::spawn_into)
+pub struct ReadyBuffer<'a> {
+    buffer: &'a mut CommandBuffer,
+}
+unsafe impl DynamicBundle for ReadyBuffer<'_> {
+    fn with_ids<T>(&self, f: impl FnOnce(&[TypeId]) -> T) -> T {
+        f(&self.buffer.ids)
+    }
+
+    fn type_info(&self) -> Vec<TypeInfo> {
+        let end =
+            self.buffer.entity[self.buffer.mark].end - self.buffer.entity[self.buffer.mark].begin;
+        self.buffer.info[0..end].iter().map(|x| x.ty_info).collect()
+    }
+
+    unsafe fn put(self, mut f: impl FnMut(*mut u8, TypeInfo)) {
+        let end =
+            self.buffer.entity[self.buffer.mark].end - self.buffer.entity[self.buffer.mark].begin;
+        for info in self.buffer.info.drain(0..end) {
```

Comment:
Repeatedly removing a prefix of a vector is wasteful, because the entire remaining length must be copied each time, resulting in quadratic overall runtime.